packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -0,0 +1,287 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert, performance } from "@fluidframework/common-utils";
+import { IConnectionDetails } from "@fluidframework/container-definitions";
+import { isOnline, OnlineStatus } from "@fluidframework/driver-utils";
+import { ProtocolOpHandler } from "@fluidframework/protocol-base";
+import { ISequencedClient } from "@fluidframework/protocol-definitions";
+import { raiseConnectedEvent, TelemetryLogger } from "@fluidframework/telemetry-utils";
+import { Audience } from "./audience";
+import { Container } from "./container";
+import { ContainerContext } from "./containerContext";
+import { DeltaManager, ReconnectMode } from "./deltaManager";
+
+export const connectEventName = "connect";
+
+export enum ConnectionState {
+    /**
+     * The document is no longer connected to the delta server
+     */
+    Disconnected,
+
+    /**
+     * The document has an inbound connection but is still pending for outbound deltas
+     */
+    Connecting,
+
+    /**
+     * The document is fully connected
+     */
+    Connected,
+}
+
+interface ILocalSequencedClient extends ISequencedClient {
+    shouldHaveLeft?: boolean;
+}
+
+export class ConnectionStateHandler {
+    private firstConnection = true;
+    private _manualReconnectInProgress = false;
+    private _messageCountAfterDisconnection: number = 0;
+    private _connectionState = ConnectionState.Disconnected;
+    private pendingClientId: string | undefined;
+    private _clientId: string | undefined;
+    private readonly _connectionTransitionTimes: number[] = [];
+
+    private get protocolHandler(): ProtocolOpHandler {
+        return this.protocolHandlerGetter();
+    }
+
+    private get context() {
+        return this.containerContextGetter();
+    }
+
+    private get audience() {
+        return this.audienceGetter();
+    }
+
+    public get connectionState(): ConnectionState {
+        return this._connectionState;
+    }
+
+    public get connected(): boolean {
+        return this.connectionState === ConnectionState.Connected;
+    }
+
+    public get clientId(): string | undefined {
+        return this._clientId;
+    }
+
+    public get messageCountAfterDisconnection() {
+        return this._messageCountAfterDisconnection;
+    }
+
+    public set messageCountAfterDisconnection(value) {
+        this._messageCountAfterDisconnection = value;
+    }
+
+    public get connectionTransitionTimes() {
+        return this._connectionTransitionTimes;
+    }
+
+    public get manualReconnectInProgress() {
+        return this._manualReconnectInProgress;
+    }
+
+    public set manualReconnectInProgress(value: boolean) {
+        this._manualReconnectInProgress = value;
+    }
+
+    constructor(
+        private readonly container: Container,
+        private readonly deltaManager: DeltaManager,
```

Comment:
Consider replacing both Container & DeltaManager with some new interfaces that have only things that this class needs. That way you have much simpler story for mocking / giving alternative implementations in testing.
Note that DeltaManager does not even need to know about this interface, as TSC will just check that implementation matches requirements (TS Duct-typing)

Reply:
Removed these arguments.